GitCommands/Git/GitGpgController.cs

```diff
@@ -0,0 +1,283 @@
+        {
+            if (revision == null)
+            {
+                throw new ArgumentNullException("revision");
+            }
+
+            TagStatus tagStatus = TagStatus.NoTag;
+
+            /* No Tag present, exit */
+            var usefulTagRefs = revision.Refs.Where(x => x.IsTag && x.IsDereference).ToList<IGitRef>();
+            if (usefulTagRefs.Count == 0)
+            {
+                return tagStatus;
+            }
+
+            return await Task.Run(() =>
+            {
+                /* More than one tag on the revision */
+                if (usefulTagRefs.Count > 1)
+                {
+                    tagStatus = TagStatus.Many;
+                }
+                else
+                {
+                    /* Raw message to be checked */
+                    string rawGpgMessage = GetTagVerificationMessage(usefulTagRefs[0], true);
+
+                    /* Look for icon to be shown */
+                    var goodSignatureMatch = GoodSignatureTagRegex.Match(rawGpgMessage);
+                    var validSignatureMatch = ValidSignatureTagRegex.Match(rawGpgMessage);
+
+                    if (goodSignatureMatch.Success && validSignatureMatch.Success)
+                    {
+                        /* It's only one good tag */
+                        tagStatus = TagStatus.OneGood;
+                    }
+                    else
+                    {
+                        Match noSignature = NoSignatureFoundTagRegex.Match(rawGpgMessage);
+                        if (noSignature.Success)
+                        {
+                            /* One tag, but not signed */
+                            tagStatus = TagStatus.TagNotSigned;
+                        }
+                        else
+                        {
+                            Match noPubKeyMatch = NoPubKeyTagRegex.Match(rawGpgMessage);
+                            if (noPubKeyMatch.Success)
+                            {
+                                /* One tag, signed, but user has not the public key */
+                                tagStatus = TagStatus.NoPubKey;
+                            }
+                            else
+                            {
+                                /* One tag, signed, any other error */
+                                tagStatus = TagStatus.OneBad;
+                            }
+                        }
+                    }
+                }
+
+                return tagStatus;
+            });
+        }
+
+        /// <summary>
+        /// Obtain the commit verification message, coming from --pretty="format:%GG" 
+        /// </summary>
+        /// <returns>Full string coming from GPG analysis on current revision.</returns>
+        public string GetCommitVerificationMessage(GitRevision revision)
+        {
+            if (revision == null)
+            {
+                throw new ArgumentNullException("revision");
+            }
+
+            var module = GetModule();
```

Comment:
This function must be provided with IGitModule

```
GetCommitVerificationMessage(IGitModule module, GitRevision revision)
```

The provided module has to be captured before the caller Task starts (line 155).

Reply:
The problem of capturing the git module before async calls spans well beyond this module. 
I propose we record it as a tech debt now and tackle it separately.